python_modules/libraries/dagster-dbt/dagster_dbt/dbt_artifacts.py

```diff
@@ -0,0 +1,152 @@
+            prepare_command: The dbt cli command to run to prepare the manifest.json
+                Default: ["parse", "--quiet"]
+            package_data_dir (Union[str, Path]):
+                A directory that will contain a copy of the dbt project, synced at build time.
+        """
+        self.project_dir = Path(project_dir)
+        if not self.project_dir.exists():
+            raise DagsterDbtProjectNotFoundError(f"project_dir {project_dir} does not exist.")
+        self._target_folder = Path(target_folder)
+        self._prepare_command = prepare_command
+        # this is ok if it doesn't exist, will get created by `prepare`
+        self._package_data_dir = Path(package_data_dir) if package_data_dir else None
+
+    @public
+    @cached_property  # to prevent repeat compilation
+    def manifest_path(self) -> Path:
+        """The path to the manifest.json, compiling the manifest first if in development or
+        ensuring it already exists if not.
+        """
+        if self._should_prepare_at_runtime():
+            return self._prepare_manifest()
+
+        p = self._base_dir.joinpath(self._target_folder, "manifest.json")
+        if not p.exists():
+            raise DagsterDbtManifestNotPreparedError(
+                f"Did not find prepared manifest.json at expected path {p}.\n"
+                "If this is an environment that is expected to prepare the manifest at run time, "
+                "set the environment variable DAGSTER_DBT_PARSE_PROJECT_ON_LOAD."
+            )
+        return p
+
+    @public
+    def prepare(self, *, quiet=False) -> None:
+        """A method that can be called as part of the deployment process to handle
+        preparing the manifest and if package_data_dir is set, handle copying
+        the dbt project to that directory.
+
+        Args:
+            quiet (bool):
+                Disable logging
+                Default: False
+        """
+        logger = None if quiet else logging.getLogger("dagster-dbt-artifacts")
+        self._prepare_manifest(logger)
+        self._handle_package_data(logger)
+        if logger:
+            logger.info("Preparation complete.")
+
+    @public
+    def get_cli_resource(self, **kwargs) -> DbtCliResource:
+        """Construct a DbtCliResource that targets the correct project directory, passing any
+        kwargs to DbtCliResource.
+        If package_data_dir is set, resolves to that outside of the development context and
+        uses project_dir when in a development context.
+        """
+        if self._should_prepare_at_runtime():
+            return DbtCliResource(project_dir=os.fspath(self.project_dir))
+
+        return DbtCliResource(
+            project_dir=os.fspath(self._base_dir),
+            **kwargs,
+        )
+
+    def _should_prepare_at_runtime(self) -> bool:
+        return (
+            # if launched via `dagster dev` cli
+            bool(os.getenv("DAGSTER_IS_DEV_CLI"))
+            or
```

Comment:
what if someone uses `dagster dev` against pre-compiled manifest.json?

Reply:
The behavior from the existing recommended setup is that if the env var is set we will still generate a new one, so thats what i went with. 

I think if you are actively iterating locally, the previous `manifest.json` will be there, so its hard to distinguish pre-compiled from development iteration. 

I think if you want to develop against a precompiled the output would be to not use this class at all. 

I could be convinced that an opt-out env var makes sense though.